### DISCUSSION TITLE
[FE210]有關於twitch的API串連

### DISCUSSION THREAD

**Discussion by wellan-H** [Lidemy/forum#150] [前端相關問題]
## 作業名稱
hw4

### 碰到的問題
不論利用舊版還是新版的API都無法繞過CORS

### 預期的行為
應該要成功在瀏覽器提取到JSON的字廖

### 嘗試過的解決方法
有看過前兩位詢問過類似問題的程式碼，但不論適用新版還是舊版都無法於瀏覽器提取JSON資料(利用GIT BASH可以)
### 相關程式碼
`function getData(){
    let cilent_id='qsjrpo1jiagbz0p1rpbn1lyaeleqh7';//使用者ID
    let access_token='f579o4vha91vapn5tvca0es096hppu';//TOKEN
    let api = 'https://api.twitch.tv/helix/streams';

    const intro= new XMLHttpRequest();
    intro.open('GET',api)
    intro.setRequestHeader('client_id',cilent_id);
    intro.setRequestHeader('Authorization','Bearer '+access_token);
    intro.send();
    intro.onload = function(e){
        const results = JSON.parse(request.response).data;
        callback(results);
      }
    
}`

### 補充說明
我嘗試了很久新版跟舊版的API，但是都可以在Git Bash提取，瀏覽器卻無法。

**Comment by outshaker** [comment]
附上參考的錯誤提示

![CORS錯誤](https://user-images.githubusercontent.com/536861/123186341-81237900-d4ca-11eb-93db-b3b0fcbce0a4.jpg)

出現非預期狀況時，開 dev tool 工具點選 Console 的欄位應該可以看見錯誤提示，根據上面的訊息提示可以發現是 `client_id` 這個 header 不在 Access-Control-Allow-Headers 清單裡面。所以調整方向是把 'client_id' 改成正確的名稱，有稍微查一下，header 大小寫在 HTTP/1.1 協議是不區分的，所以真正造成問題的是那個底線符號。

以下稍微幫你把範例碼小修一下，這樣應該就可以跑了

```html
<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Document</title>
</head>
<body>
  <script>
    function getData() {
      let cilent_id = 'qsjrpo1jiagbz0p1rpbn1lyaeleqh7' //使用者ID
      let access_token = 'f579o4vha91vapn5tvca0es096hppu' //TOKEN
      let api = 'https://api.twitch.tv/helix/streams'

      const intro = new XMLHttpRequest()
      intro.open('GET', api)
      // intro.setRequestHeader('client_id', cilent_id)
      intro.setRequestHeader('Client-Id', cilent_id) // header 的名稱要注意底線跟連字號的差異
      intro.setRequestHeader('Authorization', 'Bearer ' + access_token)
      intro.send()
      intro.onload = function(e) {
        // const results = JSON.parse(request.response).data
        const results = JSON.parse(intro.response).data // request 變數名稱不存在
        // callback(results)
        console.log(results) // callback 變數不存在
      }
    }
    getData()
  </script>
</body>
</html>
```

一些小建議：
1. 可以先試著在 dev tool 找看看有沒有錯誤訊息，通常丟 google 會找到一些線索
2. `繞過 CORS` 這件事情比較像是破解瀏覽器的安全機制，大部分會說 `符合 CORS 的要求`
3. token 會有過期問題導致你的 API 呼叫失敗，如果覺得處理這些很麻煩的話，可以暫時先用舊版的 API 迴避 token 驗證和 token 換發之類的問題

**Reply by wellan-H** [reply]
感謝回復!!
其實我一開始有丟GOOGLE，但是發現出來的回答都是要我在 response 裡加上 Access-Control-Allow-Origin header：
`Access-Control-Allow-Origin: *`
(參考至:https://shubo.io/what-is-cors/)
因為完全不知道要在哪邊加上甚麼才好，影片裡也沒看到就陷入鬼打牆中。
另外我想請問，為何我用Git Bash可以拿到資料，同樣到瀏覽器卻無法呢?

**Reply by aszx87410** [reply]
你想問的問題基本上這一系列都有提到，可以參考一下：https://blog.huli.tw/2021/02/19/cors-guide-1/

**Reply by outshaker** [reply]
我可能沒辦法像老師文章講那麼細，比較簡單的回答是 google 查到大部分的資料都是處理「後端」的設定，但你的問題是「前端」header 名稱錯誤。在這樣的情況下，你需要縮小關鍵字的範圍，才能過濾出比較接近的資料。

「git bash 可以，但是瀏覽器不行」我想按照推理的思路，你一定會猜到瀏覽器上面一定存在著「某種機制」，再加上 **CORS** 這個看起來很像關鍵字的東西，順著往下找，應該會發現一些新的關鍵字，像是 preflight, simple request, options，然後你就會覺得關鍵字越來越多，怎麼解一個問題需要知道那麼多東西

這時候偷偷分享一個小密技，你可以在關鍵字後面加上 **huli** 通常就會找到老師寫過的一些文章，可以很有效減少你撞牆的時間唷 >_^

**Reply by wellan-H** [reply]
感謝老師與大大的解答，我會再多加研讀，API算是我覺得一個很關鍵的技術點，希望能早日搞懂。
這邊有另一個問題，現在我已經透過修改後的API成功拿到資料:
![獲取的資料](https://user-images.githubusercontent.com/85210082/123263826-fa9f8380-d52b-11eb-9861-002446e0fc63.jpg)
但是卻無法透過以下的程式碼將資料導入:
![螢幕擷取畫面 2021-06-24 204511](https://user-images.githubusercontent.com/85210082/123264954-1eaf9480-d52d-11eb-846d-fdf1784bbadd.jpg)

我比較了一下新舊版本API所拿到資料的差別，發現舊版是由'streams'所包覆；新版則是'data'包覆，於是判定原影片中'const {streams}=data;'的意思是定義一個參數_streams=data{streams{}}_的意思，不料卻無法正確將遊戲名稱代入格式中，請問我是否遺漏了甚麼要素?
感謝撥冗解答。

**Reply by outshaker** [reply]
你上面寫 `$row.append(getColumn(stream));`
但下面函數宣告卻是 `function getColum(data)`
getColumn 和 getColum 名稱不一樣自然沒辦法成功呼叫

因為你附的範例程式碼是截圖不太方便丟測試，目前用肉眼方式幫你看應該是上面的錯誤
建議先試著在 dev tool 的 console 欄位那邊找看看有沒有錯誤訊息？

**Reply by wellan-H** [reply]
不好意思，我沒修改就直接貼了上來，以下是我呼叫的函式:
`
getData((err,dataid)=>{
  const data=dataid.data;
  const $row =$('.row');
  for(const stream of data) {
    $row.append(getColumn(stream));
  }
});
function getColumn(data){
  
  return`
  <div class='box'>
            <div class='box__header'>
                
            </div>
            <div class="box__banner">
              <div class="box__img">
                  
              </div>
              <div class='box__game'>${data.game_name}<br>實況主名稱</div>
            </div>
          </div>`;
}`
當我呼叫函式的時候，可以在瀏覽器dev tool 處發現得到的資料:
![螢幕擷取畫面 2021-06-25 135231](https://user-images.githubusercontent.com/85210082/123376176-9e387480-d5bc-11eb-9550-e1001d09db39.jpg)
我發現我確實可以得到資料，但卻無法將return的值確實加入到原本的HTML當中，造成畫面只剩下背景圖片。
我想請問是否是我的語法出現錯誤?但是編輯器與瀏覽器並沒有噴錯的情形，想請問是哪裡造成了問題，我不確定自己是否問問題的方式有誤，還請見諒，感謝各位。

**Reply by aszx87410** [reply]
你最主要的程式碼就是這一段：

``` js
getData((err,dataid)=>{
  const data=dataid.data;
  const $row =$('.row');
  for(const stream of data) {
    $row.append(getColumn(stream));
  }
});
```

然後結果是沒有出現在 HTML，所以要針對各個環節去 debug，依序是：

1. 在 getData 裡面加上 log，確認有呼叫到這個 callback
2. log 出 $row，確定有東西而且是預期的結果
3. log 出 data，確定有東西而且是預期的結果
4. log 出 `getColumn(stream)`，確定有東西

上面四個如果都成立，那基本上結果一定會對，東西一定會 append 到畫面上

**Reply by wellan-H** [reply]
老師你好，這邊是我加入log後的程式碼:
function getData() {
  let cilent_id = 'qsjrpo1jiagbz0p1rpbn1lyaeleqh7' //使用者ID
  let access_token = 'f579o4vha91vapn5tvca0es096hppu' //TOKEN
  let api = 'https://api.twitch.tv/helix/streams'

  const intro = new XMLHttpRequest()
  intro.open('GET', api)
  // intro.setRequestHeader('client_id', cilent_id)
  intro.setRequestHeader('Client-Id', cilent_id) // header 的名稱要注意底線跟連字號的差異
  intro.setRequestHeader('Authorization', 'Bearer ' + access_token)
  intro.send()
  intro.onload = function(e) {
    // const results = JSON.parse(request.response).data
    const results = JSON.parse(intro.response).data // request 變數名稱不存在
    // callback(results)
    console.log(results) // callback 變數不存在
    console.log("hello")//此處確認拿到資料
  }
}
getData((err,dataid)=>{
  console.log("確認是否拿到");//未印出
  const data=dataid.data;
  console.log(data);//未印出
  const $row =$('.row');
  for(const stream of data) {
    $row.append(getColumn(stream));
  }
});
經過運行後得到此情況:
![螢幕擷取畫面 2021-06-25 143239](https://user-images.githubusercontent.com/85210082/123380603-3b49dc00-d5c2-11eb-8ff8-baa161f6d3fd.jpg)
確實有將資料得到，但在呼叫時，後續設定的程式碼卻毫無作用，但把getData((err,dataid)...注解掉後，瀏覽器並沒有顯示任何資料(當然)，可是這代表我有拿到東西但後續的處理都沒有受理，是因為我呼叫的方式錯誤嗎?

**Reply by aszx87410** [reply]
你呼叫 getData 之後傳了一個函式進去，但你沒有呼叫那個函式，所以自然也就不會執行到裡面那一段的東西
如果還是不清楚為什麼，可以去找一些 callback function 的相關資料來看 👍

**Reply by wellan-H** [reply]
收到，我先去拜讀一下，等我了解了會在附上心得，謝謝老師。

**Reply by wellan-H** [reply]
老師，我去看了有關回呼函式的文章，我大致了解了回呼函式在做什麼，所以我回到了hw4的影片重新按照您的步驟寫下了這段程式碼:
![螢幕擷取畫面 2021-06-26 123038](https://user-images.githubusercontent.com/85210082/123501848-5bd56d00-d67a-11eb-8ec7-48d4e72cea02.jpg)
但是此時瀏覽器噴出這段錯誤:
![螢幕擷取畫面 2021-06-26 123146](https://user-images.githubusercontent.com/85210082/123501867-7b6c9580-d67a-11eb-923a-531fed64b62f.jpg)
有關於401的描述，我知道指的是缺乏資料所要求的必要憑證，可我根據twitch新版API的敘述:
![螢幕擷取畫面 2021-06-26 123417](https://user-images.githubusercontent.com/85210082/123501908-dacaa580-d67a-11eb-819a-6c34fe263c20.jpg)
確實的傳回了client-id以及Authorization，但卻依然拿不到資料，請問我有哪個步驟錯了嗎?

**Reply by aszx87410** [reply]
你放在網址後面的叫做 query string，最後的截圖要傳的應該是 request header，兩個是完全不同的東西喔

有些 api 支援兩個地方都可以，但是參數可能會不太一樣
